src/Transform/ONNX/ONNXDimAnalysis.cpp

```diff
@@ -0,0 +1,588 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+//===-------- ONNXDimAnalysis.cpp - ONNX Dimension Analysis ---------------===//
+//
+// Copyright 2022 The IBM Research Authors.
+//
+// =============================================================================
+//
+// This file implements an analysis on unknown dimensions in ONNX ops.
+//
+//===----------------------------------------------------------------------===//
+
+#include "mlir/IR/Matchers.h"
+#include "mlir/IR/PatternMatch.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Pass/PassManager.h"
+#include "mlir/Transforms/DialectConversion.h"
+#include "mlir/Transforms/Passes.h"
+#include "llvm/Support/Debug.h"
+
+#include "src/Dialect/ONNX/DialectBuilder.hpp"
+#include "src/Dialect/ONNX/ONNXOps.hpp"
+#include "src/Dialect/ONNX/ONNXOpsHelper.hpp"
+#include "src/Dialect/ONNX/ShapeInference/ONNXShapeHelper.hpp"
+#include "src/Pass/Passes.hpp"
+#include "src/Support/TypeUtilities.hpp"
+#include "src/Transform/ONNX/ONNXDimAnalysis.hpp"
+
+#define DEBUG_TYPE "dim_analysis"
+
+using namespace mlir;
+
+namespace {
+
+//===----------------------------------------------------------------------===//
+// Helper functions for dimension analysis.
+//===----------------------------------------------------------------------===//
+
+/// Check if two sets are interleaved or not.
+bool isInterleaved(const onnx_mlir::DimAnalysis::DimSetT &lhs,
+    const onnx_mlir::DimAnalysis::DimSetT &rhs) {
+  // Call `contains` on the smaller set for better performance.
+  if (lhs.size() < rhs.size()) {
+    for (auto &ti : lhs) {
+      if (rhs.contains(ti)) {
+        return true;
+      }
+    }
+  } else {
+    for (auto &ti : rhs) {
+      if (lhs.contains(ti)) {
+        return true;
+      }
+    }
+  }
+  return false;
+}
+
+/// Given a QuestionMarkIndexExpr representing an unknown dimension, find the
+/// same unknown dimensions in the inputs.
+void findAndAddSameDim(const onnx_mlir::QuestionmarkIndexExpr &qmOuputIE,
+    mlir::ValueRange operands, onnx_mlir::DimAnalysis::DimSetT &sameDims) {
+  int64_t outputQMValue = qmOuputIE.getQuestionmark();
+  // Cannot process if the question mark value is -1 that is a common value for
+  // every question mark.
```

Comment:
I would abstract here a bit the interface.
```
bool qmOutputIE.specificQuestionmark() 
bool qmOutputIE.sameQuestionmark(iIR)
```
whereas the first one assert if IE is not a question mark, and return true if `!= -1`,
and whereas the second one assert if either is not a question mark, and return true if their values are not `-1` and are the same (negative) value.

Reply:
I believe this comment was not resolved, at least as I see it in the code below

Reply:
Done. I created the two functions but forgot using them. Thanks!